File: DynammicProgramming/CoinChange.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] memo = new int[amount + 1];
        return cc(coins, amount, memo);
    }

    private int cc(int[] coins, int amount, int[] memo) {
        //cc(amount) = min(cc(amount - coins[i]) for each i) + 1;

        if (amount < 0) return -1;

        if (amount == 0) {
            memo[0] = 0;
            return 0;
        }

        if (memo[amount] != 0) return memo[amount];

        int res = Integer.MAX_VALUE;
        for (int coin : coins) {
            int numCoins = cc(coins, amount - coin, memo);
            if (numCoins >= 0) {
                res = Math.min(res, numCoins + 1);
            }
        }

        memo[amount] = res == Integer.MAX_VALUE ? -1 : res;
        return memo[amount];
    }
}

//================================================================================================================================

class IterativeSolution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount+1);
        dp[0] = 0;

        for (int i = 1; i <= amount; i++) {
            for (int c : coins) {
                if (i - c >= 0) {
                    dp[i] = Math.min(dp[i], 1 + dp[i - c]);
                }
            }
        }

        return (dp[amount] != amount + 1) ? dp[amount] : -1;
    }
}

File: Python/Dynammic Programming/CoinChange.py
# Recursive 
class Solution:

    def coinChange(self, coins: List[int], amount: int) -> int:
        # dp(n) = min(dp(n - k) + 1) for k in coins
        memo = [0] * (amount + 1)
        for coin in coins:
            if coin <= amount:
                memo[coin] = 1

        def dp(amount):
            nonlocal memo
            if amount < 0:
                return -1

            if amount == 0:
                return 0

            if memo[amount] != 0:
                return memo[amount]
            
            minCoins = float('inf')
            for coin in coins:
                coinNeeded = dp(amount - coin)
                if coinNeeded != -1:
                    minCoins = min(coinNeeded + 1, minCoins)
            memo[amount] = -1 if minCoins == float('inf') else minCoins
            return memo[amount]

        return dp(amount)

# Iterative 
class Solution:

    def coinChange(self, coins: List[int], amount: int) -> int:
        # dp(n) = min(dp(n - k) + 1) for k in coins
        memo = [0] * (amount + 1)
        for i in range(1, amount + 1):
            res = float("inf")
            for coin in coins:
                if (i - coin >= 0) and memo[i - coin] != -1 :
                    res = min(memo[i - coin] + 1, res)
            memo[i] = -1 if res == float("inf") else res
        return memo[amount]